barrier.go

```diff
@@ -0,0 +1,277 @@
+	return consumer.Consume()
+}
+
+func (n *BarrierNode) stopBarrierEmitter() {
+	for _, stopF := range n.barrierStopper {
+		stopF()
+	}
+}
+
+func (n *BarrierNode) NewGroup(group edge.GroupInfo, first edge.PointMeta) (edge.Receiver, error) {
+	r, stopF, err := n.newBarrier(group, first)
+	if err != nil {
+		return nil, err
+	}
+	n.barrierStopper[group.ID] = stopF
+	return edge.NewReceiverFromForwardReceiverWithStats(
+		n.outs,
+		edge.NewTimedForwardReceiver(n.timer, r),
+	), nil
+}
+
+func (n *BarrierNode) newBarrier(group edge.GroupInfo, first edge.PointMeta) (edge.ForwardReceiver, func(), error) {
+	switch {
+	case n.b.Idle != 0:
+		idleBarrier := newIdleBarrier(
+			first.Name(),
+			group,
+			n.b.Idle,
+			n.outs,
+		)
+		return idleBarrier, idleBarrier.Stop, nil
+	case n.b.Period != 0:
+		periodicBarrier := newPeriodicBarrier(
+			first.Name(),
+			group,
+			n.b.Period,
+			n.outs,
+		)
+		return periodicBarrier, periodicBarrier.Stop, nil
+	default:
+		return nil, nil, errors.New("unreachable code, barrier node should have non-zero idle or non-zero period")
+	}
+}
+
+type idleBarrier struct {
+	name  string
+	group edge.GroupInfo
+
+	idle        time.Duration
+	lastT       atomic.Value
+	wg          sync.WaitGroup
+	outs        []edge.StatsEdge
+	stopC       chan struct{}
+	resetTimerC chan struct{}
+}
+
+func newIdleBarrier(name string, group edge.GroupInfo, idle time.Duration, outs []edge.StatsEdge) *idleBarrier {
+	r := &idleBarrier{
+		name:        name,
+		group:       group,
+		idle:        idle,
+		lastT:       atomic.Value{},
+		wg:          sync.WaitGroup{},
+		outs:        outs,
+		stopC:       make(chan struct{}),
+		resetTimerC: make(chan struct{}),
+	}
+
+	r.Init()
+
+	return r
+}
+
+func (n *idleBarrier) Init() {
+	n.lastT.Store(time.Time{})
+	n.wg.Add(1)
+
+	go n.idleHandler()
+}
+
+func (n *idleBarrier) Stop() {
+	close(n.stopC)
+	n.wg.Wait()
+}
+
+func (n *idleBarrier) BeginBatch(m edge.BeginBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *idleBarrier) BatchPoint(m edge.BatchPointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) EndBatch(m edge.EndBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *idleBarrier) Barrier(m edge.BarrierMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) DeleteGroup(m edge.DeleteGroupMessage) (edge.Message, error) {
+	if m.GroupID() == n.group.ID {
+		n.Stop()
+	}
+	return m, nil
+}
+
+func (n *idleBarrier) Point(m edge.PointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+
+func (n *idleBarrier) resetTimer() {
+	n.resetTimerC <- struct{}{}
+}
+
+func (n *idleBarrier) emitBarrier() error {
+	nowT := time.Now()
```

Comment:
This should be `time.Now().UTC()` as we want to keep all time value in the UTC timezone.